### ISSUE TITLE
syscalls2 x64 and ARM64 do not support 32-bit syscalls

### ISSUE THREAD

**Issue by lacraig2** [panda-re/panda#1525]
This seems to have been a deliberate choice. I'd like to know if anyone has strong feelings about it.

If not, I'd propose we add support for it.

If people do care about it I think it'd be fine to add an argument to syscalls2 for one functionality or the other.

@LauraLMann it looks like git blame has you adding this so I want to make sure to get your input.

https://github.com/panda-re/panda/blob/1b4c51b32284bf894ad2fe941f166424edb60df0/panda/plugins/syscalls2/syscalls2.cpp#L1044-L1059

**Comment by LauraLMann** [comment]
Yes @lacraig2 , you can blame me for that.  We didn't need it for what we were doing at the time, and I couldn't think of an easy way to implement it, so I put in those warnings instead.
As long as adding it doesn't break our regression tests, I can't imagine why we would have a problem with adding 32-bit system calls in syscalls2 x64.  What do you think, @MarkMankins ?

**Comment by LauraLMann** [comment]
I have a vague recollection of thinking we'd have to support flipping back and forth between 64-bit syscalls and 32-bit syscalls in the same recording, but I may misunderstand how they really work.  (I also recall reading a lot of manuals to try to figure out what was going on.)
It sounds like you are instead thinking @lacraig2 that the use cases would only involve one or the other, not both at the same time, if you can set at plugin startup time which bitness of system calls you are interested in for a 64-bit guest. Do I understand you right?

**Comment by MarkMankins** [comment]
This would be a nice feature to have.

**Comment by AndrewQuijano** [comment]
Out of curiosity, was this the cause of LAVA crashing when injecting bugs in 32-bit binaries? Least on my end, I think the preference would be that LAVA can inject either 32-bit of 64-bit bugs. We will get LAVA to point to 64-bit bugs shortly, but I am already doing a lot with this upgrade.

**Comment by lacraig2** [comment]
> It sounds like you are instead thinking @lacraig2 that the use cases would only involve one or the other, not both at the same time, if you can set at plugin startup time which bitness of system calls you are interested in for a 64-bit guest. Do I understand you right?

I'm interested in handling any syscall that is valid for the target architecture in question in any order they are presented. On a 64-bit guest that means handling the backwards compatible 32-bit versions.

The prototypical case is actually MIPS64 which handles 3 separate ABIs by switching on the syscall number.
https://github.com/panda-re/panda/blob/dev/panda%2Fplugins%2Fsyscalls2%2Fsyscalls2.cpp#L904

It looks like ARM64 and x86_64 both overlap with their 32-bit syscall numbers so that won't work here, but we'd need something that basically carried the ABI type with the syscall request and dispatched the proper enter/return type.

**Comment by lacraig2** [comment]
Fixed
